deploy.sh

```diff
@@ -0,0 +1,56 @@
+#!/bin/bash
+
+# -------- Build Settings -------- #
+# directory to deploy
+BUILD_DIR="./dist"
+# directory on target server
+DEPLOY_DIR="./latest"
+# use to install dependencies without disruption
+TEMP_DIR="./staged"
+# server to connect to
+TARGET="hanbot@hanbot.elviswolcott.com"
+# file to run
+ENTRY="index.js"
+# name to give job (mostly for debugging on server)
+JOB_NAME="hanbot"
+
+# --------- Create .env ---------- #
+echo "\
+DISCORD_WELCOME_CHANNEL=${DISCORD_WELCOME_CHANNEL}
+DISCORD_ADMITTED_ROLE=${DISCORD_ADMITTED_ROLE}
+DISCORD_ACTIVE_ROLE=${DISCORD_ACTIVE_ROLE}
+DISCORD_TOKEN=${DISCORD_TOKEN}
+ZOOM_API_KEY=${ZOOM_API_KEY}
+ZOOM_API_SECRET=${ZOOM_API_SECRET}
+ZOOM_WEBHOOK_SCRET=${ZOOM_WEBHOOK_SECRET}
+ZOOM_MEETING_ID=${ZOOM_MEETING_ID}
+ZOOM_TIME_THRESHOLD=${ZOOM_TIME_THRESHOLD}
+ZOOM_TIME_ANNOUNCEMENT_CHANNEL=${ZOOM_TIME_ANNOUNCEMENT_CHANNEL}
+TRAVIS_BUILD_NUMBER=${TRAVIS_BUILD_NUMBER}
+TRAVIS_COMMIT=${TRAVIS_COMMIT}
+" >> ".env"
+
+
+# --------- Deploy Steps --------- #
+# remove extra files
+find "${BUILD_DIR}" -name "*.map" -type f -delete
+rm "${BUILD_DIR}/.tsBuildInfo"
+# transfer the build to the target
+rsync -r --delete-after --quiet "${TRAVIS_BUILD_DIR}/${BUILD_DIR}" "${TARGET}:${TEMP_DIR}"
+# bring over package.json & .env
+rsync -r --delete-after --quiet "${TRAVIS_BUILD_DIR}/package.json" "${TARGET}:${TEMP_DIR}/package.json"
+rsync -r --delete-after --quiet "${TRAVIS_BUILD_DIR}/.env" "${TARGET}:${TEMP_DIR}/.env"
+# restart pm2 on the target
+ssh "${TARGET}" 'bash -s' <<RESTART
+cd "${TEMP_DIR}"
+npm install --only=production
+cd ~
+pm2 stop "${JOB_NAME}"
+pm2 delete "${JOB_NAME}"
+rm -rf "${DEPLOY_DIR}"
+mv "${TEMP_DIR}" "${DEPLOY_DIR}"
+rm -rf "${TEMP_DIR}"
+cd "${DEPLOY_DIR}"
+pm2 start "${BUILD_DIR}/${ENTRY}" --name "${JOB_NAME}"
+pm2 save
```

Comment:
The way Node works, it's safe (in fact, probably better) to just overwrite the app on disk while it's running then reboot it for a zero downtime deploy. You don't need to do this jockying with temp directories.

Reply:
With the size of a normal `node_modules` directory, this sort of `mv` is going to take quite a long time (on the order of dozens of seconds).

Reply:
Although it wouldn't be a huge issue, I've run into some edge cases (I think one is when you remove a dependency in a new version) which can cause a crash. Luckily `mv`ing directories just switches the root `inode` (swaps a pointer or two around) and is virtually instantaneous.

Reply:
Three thoughts:
- I think that would only be an issue if you deleted the `node_modules` folder on the server during a deploy.
- What file system are you using with instant folder moves?
- This isn't a huge issue worth spending a lot of time on, but I'm a little worried about the complexity and non-atomicness (is that a word?) of this deploy process.

Reply:
1. IIRC it can happen when you remove a dependency between versions, causing the dependency to be removed during the install while it is still in use (I don't remember the exact conditions I had issues in, it could have been related to running native modules)
1. Pretty much file system out there I think? As long as you aren't moving it to another file system moving a directory is no different than moving a single file (as they say "everything is a file"). You can test it locally or connect to the vps and run `mv latest staged && mv staged latest` to move back and forth. Despite being ~150MB the two moves are instantaneous.
1. There's some pros and cons to doing it this way. Although it's a bit more complex, I personally prefer an opt-in vs. opt-out (for example, so that a decrypted file only used in the build process isn't transferred - as I said, not really a concern for this project).

Reply:
1. That's strange--in principle an install shouldn't delete anything. Either way, not a big deal.
2. Huh. Must be a Linux thing. On macOS (either APFS or HFS+) it needs to move every file individually.
3. Given the nature of this project (specifically, that there are a number of people who have expressed interest in contributing who don't have a lot of programming experience), I think it's valuable to make sure that contributing is as simple as possible, and adding an extra step of opting-in to deployment gets in the way of that. (And, premature optimization is the root of all email and what not.)

Reply:
I've switched to using an ignore file to exclude files instead.